ksqldb-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java

```diff
@@ -293,6 +316,186 @@ public TransientQueryMetadata executeTransientQuery(
     }
   }
 
+  public TransientQueryMetadata createStreamPullQuery(
+      final ServiceContext serviceContext,
+      final ImmutableAnalysis analysis,
+      final ConfiguredStatement<Query> statementOrig,
+      final boolean excludeTombstones
+  ) {
+
+    if (!ksqlConfig.getBoolean(KsqlConfig.KSQL_QUERY_STREAM_PULL_QUERY_ENABLED)) {
+      throw new KsqlStatementException(
+          "Pull queries on streams are disabled. To create a push query on the stream,"
+              + " add EMIT CHANGES to the end. To enable pull queries on streams, set"
+              + " the " + KsqlConfig.KSQL_QUERY_STREAM_PULL_QUERY_ENABLED + " config to 'true'.",
+          statementOrig.getStatementText()
+      );
+    }
+
+    // stream pull query overrides: start from earliest, use one  thread,
+    // and use a tight commit interval for responsiveness.
+    final ConfiguredStatement<Query> statement = statementOrig.withConfigOverrides(
+        ImmutableMap.<String, Object>builder()
+            .putAll(statementOrig.getSessionConfig().getOverrides())
+            .put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+            .put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 1)
+            .put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100)
+            .build()
+    );
+    try {
+
+      final TransientQueryMetadata transientQueryMetadata = EngineExecutor
+          .create(primaryContext, serviceContext, statement.getSessionConfig())
+          .executeTransientQuery(statement, excludeTombstones);
+
+      QueryLogger.info(
+          "Streaming stream pull query results '{}'",
+          statement.getStatementText()
+      );
+
+      return transientQueryMetadata;
+
+    } catch (final KsqlStatementException e) {
+      throw e;
+    } catch (final KsqlException e) {
+      // add the statement text to the KsqlException
+      throw new KsqlStatementException(e.getMessage(), statement.getStatementText(), e.getCause());
+    }
+  }
+
+  public void waitForStreamPullQuery(
```

Comment:
Much of the logic in KsqlEngine is somewhat orchestration rather than core logic.  Maybe we should factor out the offset fetching and checking to another class under the pull query logic?